### PR TITLE
Support EAPI 7 in kernel-2.eclass

### PR DESCRIPTION
**Pull request by dm0-**
@SoapGentoo This is the kernel eclass update with additional QA stuff in separate commits.

https://bugs.gentoo.org/702280

### CODE REVIEW

File: eclass/kernel-2.eclass
Comment (mjeveritt) [inline eclass/kernel-2.eclass:600]: We don't tend to use ranges for EAPI, as there is no guarantee (PMS) that it is, or will continue to be, numeric.

File: eclass/kernel-2.eclass
Comment (mjeveritt) [inline eclass/kernel-2.eclass:931]: Is ${ddir} assumed to have a leading slash ?

**Review by mjeveritt — APPROVED** [review]
LGTM, thanks.

File: eclass/kernel-2.eclass
Comment (thesamesam) [reply eclass/kernel-2.eclass:600]: In this case, it’s okay because we die on unknown EAPIs above anyway, and this is only a compatibility hack for older ones

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:813]: die

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:825]: die

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:854]: die

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:914]: die, die

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:954]: die

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:969]: Maybe change to a log?

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:1581]: {}

**Review by thesamesam — CHANGES_REQUESTED** [review]
Not had a chance to look for other possible improvements. Thanks for looking at this!

File: eclass/kernel-2.eclass
Comment (thesamesam) [reply eclass/kernel-2.eclass:931]: I was wondering that too :)

File: eclass/kernel-2.eclass
Comment (mjeveritt) [inline eclass/kernel-2.eclass:1439]: Perhaps keep indentation here, to indicate the continuation (I note dropping possibly redundant backslash)?

File: eclass/kernel-2.eclass
Comment (dm0-) [reply eclass/kernel-2.eclass:931]: Yes, `ddir` is assigned to the output of `kernel_header_destdir`, which always outputs a leading slash: https://github.com/gentoo/gentoo/blob/master/eclass/kernel-2.eclass#L692-L696

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:1187]: bash tests

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:1334]: bash tests

**Review by mjeveritt — COMMENTED** [review]
It can be useful to pad output to separate messages for clear visibility. This is somewhat of a cosmetic choice, and I'm not particularly for or against it personally.

File: eclass/kernel-2.eclass
Comment (mjeveritt) [reply eclass/kernel-2.eclass:600]: Hmmkay!

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:1568]: Let’s drop a bunch of ancient EAPIs given EROOT is only defined in 3+ anyway? Probably other uses too but can’t check yet

File: eclass/kernel-2.eclass
Comment (mjeveritt) [reply eclass/kernel-2.eclass:813]: `die` on the sed I assume....

File: eclass/kernel-2.eclass
Comment (thesamesam) [inline eclass/kernel-2.eclass:1635]: Inconsistent ;

File: eclass/kernel-2.eclass
Comment (mjeveritt) [reply eclass/kernel-2.eclass:825]: https://github.com/gentoo/gentoo/pull/20639#discussion_r624559875 ditto

File: eclass/kernel-2.eclass
Comment (mjeveritt) [reply eclass/kernel-2.eclass:954]: on the `find` damnit this is getting ridiculous :stuck_out_tongue_closed_eyes:

File: eclass/kernel-2.eclass
Comment (mjeveritt) [reply eclass/kernel-2.eclass:1581]: oof good catch!

File: eclass/kernel-2.eclass
Comment (mjeveritt) [reply eclass/kernel-2.eclass:1187]: `[[ ]]`

File: eclass/kernel-2.eclass
Comment (mjeveritt) [reply eclass/kernel-2.eclass:1334]: ditto `[[ ]]`

**Comment by dm0-** [comment]
I think I got everything.  There were a few more things that I thought were errors and fixed.

  - Like this not changing back to the previous directory: https://github.com/gentoo/gentoo/blob/master/eclass/kernel-2.eclass#L827-L830

  - Or these invalid array references: https://github.com/gentoo/gentoo/blob/master/eclass/kernel-2.eclass#L513

> It can be useful to pad output to separate messages for clear visibility. This is somewhat of a cosmetic choice, and I'm not particularly for or against it personally.

I'll let @SoapGentoo comment on his rejection of `echo`es.

File: eclass/kernel-2.eclass
Comment (mjeveritt) [inline eclass/kernel-2.eclass]: `${K_DEBLOB_AVAILABLE}` I think was what was meant here ..

File: eclass/kernel-2.eclass
Comment (mjeveritt) [inline eclass/kernel-2.eclass:214]: I think ordinarily we tend to prefer `"${VAR}" == "str"` syntax ...

File: eclass/kernel-2.eclass
Comment (mjeveritt) [inline eclass/kernel-2.eclass:293]: Again, for consistency's sake I think we'll stick to `""` quoting ..

File: eclass/kernel-2.eclass
Comment (dm0-) [reply eclass/kernel-2.eclass]: I did make that change.  It says this comment is on an outdated file.

File: eclass/kernel-2.eclass
Comment (mjeveritt) [inline eclass/kernel-2.eclass:418]: Not sure about arithmetic expansions in bash, so deferring to another, but any reason this might fail?

File: eclass/kernel-2.eclass
Comment (mjeveritt) [inline eclass/kernel-2.eclass:910]: ooh, there's a classic typo! would probably preserve the space after `&>` for readability. </nitpick>

File: eclass/kernel-2.eclass
Comment (mjeveritt) [reply eclass/kernel-2.eclass]: just noticed that ! :)

**Comment by mjeveritt** [comment]
@mpagano @Whissi @aliceinwire for final thoughts before ML review, etc

File: eclass/kernel-2.eclass
Comment (telans) [reply eclass/kernel-2.eclass:293]: Bash doesn't do any word splitting within `[[ ]]`. Quotes don't matter if they're there but it is redundant

File: eclass/kernel-2.eclass
Comment (a17r) [reply eclass/kernel-2.eclass:214]: In general we tell people to drop quotation inside ``[[ ]]``.

File: eclass/kernel-2.eclass
Comment (dm0-) [reply eclass/kernel-2.eclass:910]: Nearly all other instances of redirection in the file have no space, so I'd prefer to omit it for consistency rather than rewrite all the other lines, unless that is the coding standard.

**Comment by dm0-** [comment]
> for final thoughts before ML review, etc

Note the EAPI 7 patch has not changed since [the last time I sent this to the mailing list](https://archives.gentoo.org/gentoo-dev/message/a032ddd8d54fb0d0797f855a69a4e612), which was never acknowledged.  This PR just has additional cleanup commits.

**Comment by dm0-** [comment]
@SoapGentoo I had to revert that `emake` for reading variables that I mentioned earlier since it's apparently expected to fail sometimes and should not kill the build.

**Comment by SoapGentoo** [comment]
@Whissi I'd like to have a hard deadline for this, and of not, we'll merge this

**Comment by dm0-** [comment]
@mpagano I applied your patch from the bug here to run it through CI.  If you're a maintainer of this GitHub project, you should be able to modify the PR branch if needed.

**Pull request closed without merging**
